backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/IrModuleDeserializer.kt

```diff
@@ -0,0 +1,1130 @@
+            "private_to_this" -> Visibilities.PRIVATE_TO_THIS
+            "protected" -> Visibilities.PROTECTED
+            "internal" -> Visibilities.INTERNAL
+            "invisible_fake" -> Visibilities.INVISIBLE_FAKE // TODO: eventually we should not serialize fake overrides, so this will be gone.
+            "local" -> Visibilities.LOCAL
+            else -> error("Unexpected visibility value: $value")
+        }
+    }
+
+    private fun deserializeIrConstructor(
+        proto: KonanIr.IrConstructor,
+        start: Int,
+        end: Int,
+        origin: IrDeclarationOrigin
+    ): IrConstructor {
+        val symbol = deserializeIrSymbol(proto.symbol) as IrConstructorSymbol
+
+        val constructor = symbolTable.declareConstructor(UNDEFINED_OFFSET, UNDEFINED_OFFSET, irrelevantOrigin,
+            symbol.descriptor, {
+                IrConstructorImpl(
+                    start, end, origin,
+                    it,
+                    Name.identifier(proto.base.name),
+                    deserializeVisibility(proto.base.visibility),
+                    deserializeIrType(proto.base.returnType),
+                    proto.base.isInline,
+                    proto.base.isExternal,
+                    proto.isPrimary
+                )
+
+            })
+
+        deserializeIrFunctionBase(proto.base, constructor as IrFunctionBase, start, end, origin)
+        return constructor
+    }
+
+    private fun deserializeIrField(proto: KonanIr.IrField, start: Int, end: Int, origin: IrDeclarationOrigin): IrField {
+        val symbol = deserializeIrSymbol(proto.symbol) as IrFieldSymbol
+        val field = IrFieldImpl(
+            start, end, origin,
+            symbol,
+            Name.identifier(proto.name),
+            deserializeIrType(proto.type),
+            deserializeVisibility(proto.visibility),
+            proto.isFinal,
+            proto.isExternal,
+            proto.isStatic
+        )
+        val initializer = if (proto.hasInitializer()) deserializeExpression(proto.initializer) else null
+        field.initializer = initializer?.let { IrExpressionBodyImpl(it) }
+
+        return field
+    }
+
+    private fun deserializeModality(modality: KonanIr.ModalityKind) = when (modality) {
+        KonanIr.ModalityKind.OPEN_MODALITY -> Modality.OPEN
+        KonanIr.ModalityKind.SEALED_MODALITY -> Modality.SEALED
+        KonanIr.ModalityKind.FINAL_MODALITY -> Modality.FINAL
+        KonanIr.ModalityKind.ABSTRACT_MODALITY -> Modality.ABSTRACT
+    }
+
+    private fun deserializeIrProperty(
+        proto: KonanIr.IrProperty,
+        start: Int,
+        end: Int,
+        origin: IrDeclarationOrigin
+    ): IrProperty {
+
+        // We start with the field to get its descriptor
+        val backingField = if (proto.hasBackingField()) {
+            deserializeIrField(proto.backingField, start, end, origin)
+        } else null
+
+        val descriptor = backingField?.descriptor
```

Comment:
Is this correct?
Consider delegated property. It has backing field but the property itself has different descriptor.
At least it was the case before, not sure if I describe current state properly.

Reply:
That was incorrect. Removed.